imagedephi/redact.py

```diff
@@ -1,74 +1,166 @@
-                    redact_tiff_tags(sub_ifds, tags_to_redact)
-
-
-def redact_one_image(tiff_info: TiffInfo, output_path: Path) -> None:
-    ifds = tiff_info["ifds"]
-    tags_to_redact = get_tags_to_redact()
-    redact_tiff_tags(ifds, tags_to_redact)
-    tifftools.write_tiff(tiff_info, output_path)
+class TiffMetadataRedactionPlan:
+    """
+    Represents a plan of action for redacting metadata from TIFF images.
+
+    The plan can be used for reporting to end users what steps will be made to redact their TIFF
+    images, and also executing the plan.
+    """
+
+    redaction_steps: dict[int, tuple[TiffMetadataRule, RuleSource]]
+    no_match_tags: list[tifftools.TiffTag]
+    image_data: TiffInfo
+    base_rules: list[TiffMetadataRule]
+    override_rules: list[TiffMetadataRule]
+
+    @staticmethod
+    def _iter_tiff_tag_entries(
+        ifds: list[IFD],
+    ) -> Generator[tuple[tifftools.TiffTag, IFD], None, None]:
+        for ifd in ifds:
+            for tag_id, entry in sorted(ifd["tags"].items()):
+                tag: tifftools.TiffTag = tifftools.constants.get_or_create_tag(
+                    tag_id, datatype=tifftools.Datatype[entry["datatype"]]
+                )
+                if not tag.isIFD():
+                    yield tag, ifd
+                else:
+                    # entry['ifds'] contains a list of lists
+                    # see tifftools.read_tiff
+                    for sub_ifds in entry.get("ifds", []):
+                        yield from TiffMetadataRedactionPlan._iter_tiff_tag_entries(sub_ifds)
+
+    def _add_tag_to_plan(self, tag: tifftools.TiffTag) -> None:
+        """Determine how to handle a given tag."""
+        for rule in self.override_rules:
+            if rule.is_match(tag):
+                self.redaction_steps[tag.value] = (rule, RuleSource.OVERRIDE)
+                return
+        for rule in self.base_rules:
+            if rule.is_match(tag):
+                self.redaction_steps[tag.value] = (rule, RuleSource.BASE)
+                return
+        self.no_match_tags.append(tag)
+
+    def _build_redaction_steps(self, ifds: list[IFD]) -> None:
+        for tag, _ in self._iter_tiff_tag_entries(ifds):
+            self._add_tag_to_plan(tag)
+
+    def __init__(
+        self, tiff_info: TiffInfo, base_rules: RuleSet, override_rules: RuleSet | None
+    ) -> None:
+        self.redaction_steps = {}
+        self.no_match_tags = []
+        self.image_data = tiff_info
+        self.base_rules = [
+            rule for rule in base_rules.rules[RuleFormat.TIFF] if isinstance(rule, TiffMetadataRule)
+        ]
+        if override_rules is None:
+            self.override_rules = []
+        else:
+            self.override_rules = [
+                rule
+                for rule in override_rules.rules[RuleFormat.TIFF]
+                if isinstance(rule, TiffMetadataRule)
+            ]
+        self._build_redaction_steps(self.image_data["ifds"])
```

Comment:
It seems like you could inline `_build_redaction_steps` without making things much more complicated here. If this class ends up being that difficult to construct, I think we should try to make it simple.

Reply:
Inlined both this function and `_redact_image` in 95f43bf. The use of the generator to iterate over IFDs made these functions perhaps too small to justify being functions themselves.